ueberdosis/tiptap
# feat(core): NodePositions

## Please describe your changes

This PR adds a new class to the core library called `NodePosition`. This is a wrapper around ProseMirror's `ResolvedPos` aiming to make handling content & content positions easier.

With this class, you could do stuff like this:

```js
const { $anchor } = editor.state.selection
const nodePos = new NodePosition($anchor)

// Delete the node before the currently anchored node
editor
  .chain()
  .deleteRange({ from: nodePos.before.from, to: nodePos.before.to })
  .focus()
  .run()
```

or

```js
const { $anchor } = editor.state.selection
const nodePos = new NodePosition($anchor)

// Add some text behind the current list
const bulletListParent = nodePos.getParentByType('bulletList')
if (!bulletListParent) {
  return false
}

editor
  .chain()
  .insertContentAt(bulletListParent.to + 1, 'Hello World!')
  .focus()
  .run()
```

The `NodePosition` class has the following values & getters:

- `NodePosition.node` - The node found at the current ResolvedPos
- `NodePosition.$pos` - The current resolved position
- `NodePosition.doc` - The parent document node
- `NodePosition.depth` - The depth of the current NodePosition
- `NodePosition.name` - The node typename found at the NodePosition
- `NodePosition.from` - The start position of the NodePosition
- `NodePosition.to` - The end position of the NodePosition
- `NodePosition.before` - The NodePosition before this NodePosition
- `NodePosition.after` - The NodePosition after this NodePosition
- `NodePosition.parent` - The parent NodePosition for this NodePosition
  - This could be used for handling nested elements
- `NodePosition.children` - Returns an array of direct children if there are any
- `NodePosition.firstChild` - Returns the first child if there is one
- `NodePosition.lastChild` - Returns the last child if there is one
- `NodePosition.range` - The range of this NodePosition
- `NodePosition.createNodeRange` - Creates a new NodeRange for this NodePosition
- `getParentByType(typeOrName)` - Finds a specific Node type as parent and returns it's NodePosition
- `hasParentByType(typeOrName)` - Checks if a specific Node type as parent exists
- `NodePosition.getChildrenOnDepth(targetDepth: number)` - Returns all children on the given depth of this NodePosition
- `NodePosition.getDeepChildren()` - Returns ALL children inside this NodePosition (including deep childs like `paragraph` and `text` nodes)

It also has two static functions called the same as the dynamic functions for parents.
I'm currently investigating how I can make the handling even better (for example also going through children, finding specific child NodePositions, etc.)

In the end something like this is now possible:

```js
// Get the position of the parent of this parents predecessor
const targetPosition = myNodePos.parent.before.parent.from
```

I'm working on this to make my own work on some of our issues like List behavior #3819 or other nested nodes.

## How did you accomplish your changes

Take a look at the class:
https://github.com/ueberdosis/tiptap/compare/feature/node-positions?expand=1#diff-2d825cb76dab95991ece6347a3abec26ba8185c10374d30329656622976126aa

## How have you tested your changes

I created a local demo page showing and tracking all NodePositions important to me with Decorations & I console logged them on update and selectionChange to make sure the values come out correctly.

## How can we verify your changes

1. Create a local demo from this branch
2. Create an editor
3. Inside the onUpdate and onSelectionChange hooks add this:

```js
const { $anchor } = editor.state.selection
console.log(new NodePosition($anchor))
```

4. Check out the console output and try different setups

## Remarks

This is still early – I'll also need to see how well it performs but since most of the more complex things only happen when the getters are called I think it's already pretty good.

## Checklist

- [x] Add documentation where needed
